src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/P11HKDF.java

```diff
@@ -227,6 +234,26 @@ private <T> T derive(String alg, AlgorithmParameterSpec derivationSpec,
         }
     }
 
+    private long getDerivedKeyType(P11SecretKeyFactory.KeyInfo ki, String alg)
+            throws InvalidAlgorithmParameterException {
+        switch ((int) ki.keyType) {
+            case (int) CKK_DES, (int) CKK_DES3, (int) CKK_AES, (int) CKK_RC4,
+                    (int) CKK_BLOWFISH, (int) CKK_CHACHA20,
+                    (int) CKK_GENERIC_SECRET -> {
+                if (ki.keyType != CKK_GENERIC_SECRET ||
+                        alg.equalsIgnoreCase("Generic")) {
+                    return ki.keyType;
+                }
+            }
+            case (int) PCKK_TLSPREMASTER, (int) PCKK_TLSRSAPREMASTER,
+                    (int) PCKK_TLSMASTER -> {
+                return CKK_GENERIC_SECRET;
+            }
```

Comment:
It's easier to troubleshoot to add a default case and not let it fall through to the exception on line 253? It's possible that P11SecretKeyFactory is enhanced with more KeyInfo, but the newly added keyType is not added here. Lumping different causes into the same exception may be harder to debug.

Reply:
The exception informs the algorithm, and we know that the algorithm was found in the map because, otherwise, we would have not been able to get the `KeyInfo ki` received by parameter. I can add two separate exceptions if you want, but should not make much of a difference because the reason for the exception is the same: the algorithm is not valid for derivation, regardless if its underlying mechanism is CKK_GENERIC_SECRET or something else.